File: scala/scala-impl/src/org/jetbrains/plugins/scala/lang/psi/impl/base/ScLiteralEscaper.java
package org.jetbrains.plugins.scala.lang.psi.impl.base;

import com.intellij.openapi.util.ProperTextRange;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.LiteralTextEscaper;
import com.intellij.psi.impl.source.tree.java.PsiLiteralExpressionImpl;
import org.jetbrains.annotations.NotNull;

/**
 * Pavel Fatin
 */

public class ScLiteralEscaper extends LiteralTextEscaper<ScLiteralImpl> {
  private int[] outSourceOffsets;

  public ScLiteralEscaper(final ScLiteralImpl literal) {
    super(literal);
  }

  public boolean decode(@NotNull final TextRange rangeInsideHost, @NotNull StringBuilder outChars) {
    ProperTextRange.assertProperRange(rangeInsideHost);
    String subText = rangeInsideHost.substring(myHost.getText());
    outSourceOffsets = new int[subText.length() + 1];
    return PsiLiteralExpressionImpl.parseStringCharacters(subText, outChars, outSourceOffsets);
  }

  public int getOffsetInHost(int offsetInDecoded, @NotNull final TextRange rangeInsideHost) {
    int result = offsetInDecoded < outSourceOffsets.length ? outSourceOffsets[offsetInDecoded] : -1;
    if (result == -1) return -1;
    return (result <= rangeInsideHost.getLength() ? result : rangeInsideHost.getLength()) + rangeInsideHost.getStartOffset();
  }

  public boolean isOneLine() {
    final Object value = myHost.getValue();
    return value instanceof String && ((String)value).indexOf('\n') < 0;
  }
}

File: scala/scala-impl/src/org/jetbrains/plugins/scala/lang/psi/impl/base/ScLiteralEscaper.scala
package org.jetbrains.plugins.scala.lang.psi.impl.base

import com.intellij.openapi.util.TextRange
import com.intellij.psi.LiteralTextEscaper
import com.intellij.psi.impl.source.tree.java.PsiLiteralExpressionImpl
import org.jetbrains.plugins.scala.lang.psi.api.base.literals.ScStringLiteral

class ScLiteralEscaper(val literal: ScStringLiteral) extends LiteralTextEscaper[ScStringLiteral](literal) {
  private var outSourceOffsets: Array[Int] = _

  override def decode(rangeInsideHost: TextRange, outChars: java.lang.StringBuilder): Boolean = {
    TextRange.assertProperRange(rangeInsideHost)
    val subText = rangeInsideHost.substring(myHost.getText)
    outSourceOffsets = new Array[Int](subText.length + 1)
    PsiLiteralExpressionImpl.parseStringCharacters(subText, outChars, outSourceOffsets)
  }

  override def getOffsetInHost(offsetInDecoded: Int, rangeInsideHost: TextRange): Int = {
    val result =
      if (offsetInDecoded < outSourceOffsets.length) outSourceOffsets(offsetInDecoded)
      else -1

    if (result == -1) {
      -1
    } else {
      (if (result <= rangeInsideHost.getLength) result else rangeInsideHost.getLength) +
        rangeInsideHost.getStartOffset
    }
  }

  override def isOneLine: Boolean = {
    myHost.getValue match {
      case str: String => str.indexOf('\n') < 0
      case _ => false
    }
  }
}